doc/api/stream.md

```diff
@@ -545,210 +762,59 @@ myReader.on('readable', () => {
-the callback is attached as a listener on the [`'finish'`][] event.
-
-Calling [`stream.write()`][stream-write] after calling
-[`stream.end()`][stream-end] will raise an error.
-
-```js
-// write 'hello, ' and then end with 'world!'
-var file = fs.createWriteStream('example.txt');
-file.write('hello, ');
-file.end('world!');
-// writing more now is not allowed!
-```
-
-#### writable.setDefaultEncoding(encoding)
-
-* `encoding` {String} The new default encoding
-* Return: `this`
-
-Sets the default encoding for a writable stream.
-
-#### writable.uncork()
-
-Flush all data, buffered since [`stream.cork()`][] call.
-
-#### writable.write(chunk[, encoding][, callback])
-
-* `chunk` {String|Buffer} The data to write
-* `encoding` {String} The encoding, if `chunk` is a String
-* `callback` {Function} Callback for when this chunk of data is flushed
-* Returns: {Boolean} `true` if the data was handled completely.
 
-This method writes some data to the underlying system, and calls the
-supplied callback once the data has been fully handled.  If an error
-occurs, the callback may or may not be called with the error as its
-first argument.  To detect write errors, listen for the `'error'` event.
+#### Class: stream.Transform
 
-The return value indicates if you should continue writing right now.
-If the data had to be buffered internally, then it will return
-`false`. Otherwise, it will return `true`.
+<!--type=class-->
+
+Transform streams are [Duplex][] streams where the output is in some way
+computed from the input. Like all [Duplex][] streams, Transform streams
+implement both the [Readable][] and [Writable][] interfaces.
+
+Examples of Transform streams include:
 
-This return value is strictly advisory. You MAY continue to write,
-even if it returns `false`. However, writes will be buffered in
-memory, so it is best not to do this excessively. Instead, wait for
-the [`'drain'`][] event before writing more data.
+* [zlib streams][zlib]
+* [crypto streams][crypto]
 
 
 ## API for Stream Implementors
 
 <!--type=misc-->
 
-To implement any sort of stream, the pattern is the same:
+The `stream` module API has been designed to make it possible to easily
+implement streams using JavaScript's prototypical inheritance model.
+
+First, a stream developer would declare a new JavaScript class that extends one
+of the four basic stream classes (`stream.Writable`, `stream.Readable`,
+`stream.Duplex`, or `stream.Transform`), making sure the call the appropriate
+parent class constructor:
 
-1. Extend the appropriate parent class in your own subclass via the `extends`
-   keyword.
-2. Call the appropriate parent class constructor in your constructor,
-   to be sure that the internal mechanisms are set up properly.
-3. Implement one or more specific methods, as detailed below.
+```js
+const Writable = require('stream').Writable;
+
+class MyWritable extends Writable {
+  constructor(options) {
+    super(options);
+  }
+}
```

Comment:
We've sort-of-kind-of optimized the API for the following usage with `options.write` / `options.transform` / et al:

```js
function createWritable (opts) {
  return Writable({
    write (chunk, enc, cb) {
    }
  })
}
```